File: examples/ts/multi-method/MultiMethod.test.ts
import { describe, it, expect } from 'vitest';
import { readFileSync } from 'node:fs';
import { join, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';
import { TestContract, ALICE, BOB, signTestMessage } from 'runar-testing';

const __dirname = dirname(fileURLToPath(import.meta.url));
const source = readFileSync(join(__dirname, 'MultiMethod.runar.ts'), 'utf8');

const ALICE_SIG = signTestMessage(ALICE.privKey);
const BOB_SIG = signTestMessage(BOB.privKey);

describe('MultiMethod', () => {
  it('compiles via TestContract.fromSource', () => {
    const c = TestContract.fromSource(source, { owner: ALICE.pubKey, backup: BOB.pubKey });
    expect(c.state.owner).toBe(ALICE.pubKey);
    expect(c.state.backup).toBe(BOB.pubKey);
  });

  it('spendWithOwner succeeds when threshold is met and owner signs', () => {
    const c = TestContract.fromSource(source, { owner: ALICE.pubKey, backup: BOB.pubKey });
    // amount * 2 + 1 must be > 10, so amount >= 5.
    const result = c.call('spendWithOwner', { sig: ALICE_SIG, amount: 6n });
    expect(result.success).toBe(true);
  });

  it('spendWithOwner rejects when the threshold is not met', () => {
    const c = TestContract.fromSource(source, { owner: ALICE.pubKey, backup: BOB.pubKey });
    const result = c.call('spendWithOwner', { sig: ALICE_SIG, amount: 1n });
    expect(result.success).toBe(false);
  });

  it('spendWithBackup succeeds with the backup signature', () => {
    const c = TestContract.fromSource(source, { owner: ALICE.pubKey, backup: BOB.pubKey });
    const result = c.call('spendWithBackup', { sig: BOB_SIG });
    expect(result.success).toBe(true);
  });
});

File: examples/zig/multi-method/MultiMethod_test.zig
const std = @import("std");
const root = @import("../examples_test.zig");
const runar = @import("runar");
const MultiMethod = @import("MultiMethod.runar.zig").MultiMethod;

fn contractPath(comptime basename: []const u8) []const u8 {
    return "multi-method/" ++ basename;
}

fn runCompileChecks(comptime basename: []const u8) !void {
    try root.runar.compileCheckSource(std.testing.allocator, @embedFile(basename), basename);
    try root.runar.compileCheckFile(std.testing.allocator, contractPath(basename));
}

test "compile-check MultiMethod.runar.zig" {
    try runCompileChecks("MultiMethod.runar.zig");
}

test "MultiMethod.spendWithOwner succeeds when the owner signs and threshold is met" {
    const contract = MultiMethod.init(runar.ALICE.pubKey, runar.BOB.pubKey);
    contract.spendWithOwner(runar.signTestMessage(runar.ALICE), 6); // 6 * 2 + 1 = 13 > 10
}

test "MultiMethod.spendWithBackup succeeds when the backup signs" {
    const contract = MultiMethod.init(runar.ALICE.pubKey, runar.BOB.pubKey);
    contract.spendWithBackup(runar.signTestMessage(runar.BOB));
}
